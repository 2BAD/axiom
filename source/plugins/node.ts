import type { Linter } from 'eslint'
import plugin from 'eslint-plugin-n'

export const config: Linter.Config[] = [
  {
    ...plugin.configs['flat/recommended-module'],
    name: 'eslint/plugin/node/recommended-module'
  },
  {
    name: 'axiom/plugin/node/overrides',
    rules: {
      'n/prefer-node-protocol': 'error',
      'n/no-missing-import': 'off'
    }
  }
]

export default config
