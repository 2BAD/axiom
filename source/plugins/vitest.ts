import plugin from '@vitest/eslint-plugin'
import type { Linter } from 'eslint'

export const config: Linter.FlatConfig[] = [
  {
    name: '2bad/eslint-plugin-vitest',
    ...plugin.configs.all,
    ...plugin.configs.env,
    files: ['**/*.test.ts'],
    // https://vitest.dev/guide/testing-types
    settings: {
      vitest: {
        typecheck: true
      }
    },
    languageOptions: {
      globals: {
        ...plugin.environments.env.globals
      }
    },
    rules: {
      // this rules are not part of 'all' preset for some reason
      ...plugin.configs.recommended.rules,
      'vitest/max-nested-describe': ['error', { max: 3 }],
      'vitest/no-hooks': ['error', { allow: ['afterEach'] }]
    }
  }
]

export default config
