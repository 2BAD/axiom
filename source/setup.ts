import { includeIgnoreFile } from '@eslint/compat'
import type { Linter } from 'eslint'
import { resolve } from 'node:path'

export const setup = (dirname: string): Linter.FlatConfig[] => {
  return [
    includeIgnoreFile(resolve(dirname, '.gitignore')),
    {
      name: '2bad/setup/linter-options',
      linterOptions: {
        reportUnusedDisableDirectives: 'error'
      }
    },
    {
      name: '2bad/setup/language-options',
      languageOptions: {
        parserOptions: {
          projectService: true,
          tsconfigRootDir: import.meta.dirname
        }
      }
    },
    {
      name: '2bad/setup/file-extension',
      files: ['**/*.ts', '**/*.cts', '**.*.mts']
    }
  ]
}
