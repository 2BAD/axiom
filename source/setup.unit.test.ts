import { includeIgnoreFile } from '@eslint/compat'
import { resolve } from 'node:path'
import { describe, expect, it, vi } from 'vitest'
import setup from './setup.ts'

const ignores = ['**/logs', '**/tmp', '**/node_modules', '**/build', '**/coverage', '**/.env']

vi.mock('@eslint/compat', () => ({
  includeIgnoreFile: vi.fn((_) => ({
    name: 'Imported .gitignore patterns',
    ignores
  }))
}))

describe('setup', () => {
  it('should return an array of FlatConfig objects', () => {
    const dirname = '/some/dir'
    const expected = [
      {
        name: 'Imported .gitignore patterns',
        ignores
      },
      {
        name: '2bad/setup/linter-options',
        linterOptions: {
          reportUnusedDisableDirectives: 'error'
        }
      },
      {
        name: '2bad/setup/file-extension',
        files: ['**/*.ts', '**/*.cts', '**.*.mts']
      }
    ]

    const result = setup(dirname)

    expect(result).toEqual(expected)
  })

  it('should correctly resolve the .gitignore path', () => {
    const dirname = '/another/dir'
    const result = setup(dirname)
    const resolvedPath = resolve(dirname, '.gitignore')

    expect(includeIgnoreFile).toHaveBeenCalledWith(resolvedPath)
    expect(result[0]).toEqual({
      name: 'Imported .gitignore patterns',
      ignores
    })
  })

  it('should include the linter-options object in the result', () => {
    const dirname = '/any/dir'
    const result = setup(dirname)
    const linterOptionsConfig = result.find((config) => config.name === '2bad/setup/linter-options')

    expect(linterOptionsConfig).toBeDefined()
    expect(linterOptionsConfig?.linterOptions).toEqual({
      reportUnusedDisableDirectives: 'error'
    })
  })

  it('should include the file-extension object in the result', () => {
    const dirname = '/any/dir'
    const result = setup(dirname)
    const fileExtensionConfig = result.find((config) => config.name === '2bad/setup/file-extension')

    expect(fileExtensionConfig).toBeDefined()
    expect(fileExtensionConfig?.files).toEqual(['**/*.ts', '**/*.cts', '**.*.mts'])
  })
})
