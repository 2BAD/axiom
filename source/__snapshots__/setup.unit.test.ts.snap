// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setup > should return an array of FlatConfig objects 1`] = `
[
  {
    "ignores": [
      "**/logs",
      "**/tmp",
      "**/node_modules",
      "**/build",
      "**/coverage",
      "**/.env",
    ],
    "name": "Imported .gitignore patterns",
  },
  {
    "linterOptions": {
      "reportUnusedDisableDirectives": "error",
    },
    "name": "2bad/setup/linter-options",
  },
  {
    "languageOptions": {
      "parserOptions": {
        "projectService": true,
        "tsconfigRootDir": "/home/projects/2BAD/eslint-config/source",
      },
    },
    "name": "2bad/setup/language-options",
  },
  {
    "files": [
      "**/*.ts",
      "**/*.cts",
      "**.*.mts",
    ],
    "name": "2bad/setup/file-extension",
  },
]
`;
